File: Python/7py 2/prefix_suffix.py
"""
|--------------------------------------------------------------------------
| String prefix and suffix
|--------------------------------------------------------------------------
|
"""

def solve(s):
   x = 0
   for i in range(len(s) // 2 + 1):
      if s[:i] == s[-i:]: x = i
   return x


solve=lambda s:len(__import__('re').match(r'^(.*).*\1$',s).group(1))


def solve(st):
   i = len(st)//2
   while i and not st.endswith(st[0:i]): i -= 1
   return i


print(solve("abcd"),0)
print(solve("abcda"),1)
print(solve("abcdabc"),3)
print(solve("abcabc"),3)
print(solve("abcabca"),1)
print(solve("aaaaa"),2)
print(solve("aaaa"),2)
print(solve("aaa"),1)
print(solve("aa"),1)
print(solve("a"),0)

File: JavaScript/7js 3/prefix_suffix.js
/*
|--------------------------------------------------------------------------
| String prefix and suffix
|--------------------------------------------------------------------------
|
*/

function solve(s) {
   let x = 0;
   for (let i = 0; i <= s.length / 2; i++) {
      x = s.substring(0, i) == s.substring(s.length - i) ? i : x;
   }
   return x;
}


function solve(s) {
   return s.match(/^(.*).*\1$/)[1].length;
}


function solve(s, i = s.length >> 1) {
   while (i && s.slice(0, i) != s.slice(-i)) --i;
   return i;
}


console.log(solve("abcd"), 0);
console.log(solve("abcda"), 1);
console.log(solve("abcdabc"), 3);
console.log(solve("abcabc"), 3);
console.log(solve("abcabca"), 1);
console.log(solve("aaaa"), 2);
console.log(solve("aa"), 1);
console.log(solve("a"), 0);